### PR TITLE
Fix ModuleListPresenter tests

### PR DESCRIPTION
**Pull request by armstrongnate**
[ignore-commit-lint]

The main issue here was that `UseCase.fetch` callbacks were being
completed in tests that did not initiate them.

- TestA kicks off a request for courses and a request for modules
- TestA waits until courses load then the test would finish
- TestB starts
- TestB mocks modules request
- TestB requests modules
- TestA's request for modules would finish
- **TestA's modules would get written to TestB's db**
- TestB's request for modules would finish
- TestB's modules would get written to TestB's db
- TestB would fail because too many modules were written

This is due to UseCases retaining themselves until the callbacks
complete and not having a way to cancel a UseCase..

We should _probably_ support canceling at some point but in the
meantime, ensuring that we write to the same database that initiated the
`fetch` fixes test issues since we re-create the db for every test.

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please run `yarn update-headers`. The following do not have the correct license header: 
* Core/CoreTests/Extensions/DispatchExtensionsTests.swift
  </td>
    </tr>
  </tbody>
</table>



Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Core** | 94.13% | 94.14% | -0%
**Student** | 93.36% | 93.36% | 0%
**Teacher** | 100% | 100% | 0%
**React Native** | 92.54% | 92.54% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Core/Core/Extensions/DispatchExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/// Performs `closure` on the main thread
+///
+/// If the executing thread is already the main thread, the `closure`
+/// is executed immediately.
+public func performUIUpdate(using closure: @escaping () -> Void) {
+    // If we are already on the main thread, execute the closure directly
+    if Thread.isMainThread {
+        closure()
+    } else {
+        DispatchQueue.main.async(execute: closure)
```
Comment (bluwave) [inline Core/Core/Extensions/DispatchExtensions.swift]: clever

File: Core/Core/Store/UseCase.swift [hunk 2]
```diff
@@ -63,6 +63,9 @@ extension UseCase {
     }
 
     public func fetch(environment: AppEnvironment = .shared, force: Bool = false, _ callback: @escaping RequestCallback) {
+        // Make sure we write to the database that initiated this request
+        let database = environment.database
```
Comment (bluwave) [inline Core/Core/Store/UseCase.swift]: i'm lost here, was another environment being used than the one passed in?

**Comment by bluwave** [comment]
would the `tearDown` method in `CoreTestCase` be of any help to reset useCases, callbacks, or database?

**Comment by shichongrui** [comment]
Looks like there are still some flaky tests.

File: Core/Core/Store/UseCase.swift (hunk 2 above)
Comment (armstrongnate) [reply Core/Core/Store/UseCase.swift]: The `environment` doesn't change but `environment.database` does change between the time the `fetch` starts and completes.

**Comment by armstrongnate** [comment]
@bluwave Ideally we would cancel the use cases on tearDown but we don't have a way to do that right now.

@shichongrui Thanks, I'll look at the flaky test.

I'm glad this pr is exposing a bunch of flakiness.

**Pull request merged**